Add tools section to Boost library views and templates

julioest opened the pull request:
- Introduced a new TOOLS constant in constants.py containing various Boost tools with descriptions and URLs.
- Implemented a get_tools function in utils.py to retrieve and sort the tools.
- Updated LibraryListBase to include tools in the context data for library views.
- Enhanced categorized_list.html, grid_list.html, and vertical_list.html templates to display tools alongside libraries, with appropriate descriptions and formatting.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added a “Tools” section to library listings across grid, vertical, and categorized views.
  * Included Boost-related tools with descriptions and external links, with version-aware documentation routing when available.
  * Added responsive layouts for tool cards/lists (desktop and mobile).
* **Bug Fixes**
  * Refined categorized empty states so tools-only categories display correctly.
* **Style**
  * Improved spacing in library view and filter controls.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

julioest committed:
WIP: Add tools section to Boost library views and templates

- Introduced a new TOOLS constant in constants.py containing various Boost tools with descriptions and URLs.
- Implemented a get_tools function in utils.py to retrieve and sort the tools.
- Updated LibraryListBase to include tools in the context data for library views.
- Enhanced categorized_list.html, grid_list.html, and vertical_list.html templates to display tools alongside libraries, with appropriate descriptions and formatting.

julioest committed:
Enhance tools functionality in Boost library

- Added a new tool "boostlook" to the TOOLS constant in constants.py with its description and URL.
- Modified the get_tools function in utils.py to accept an optional version parameter, allowing for dynamic URL generation for version-specific tools.
- Updated LibraryListBase and LibraryCategorized views to pass the selected version to get_tools, ensuring correct URL handling in the context data.

julioest committed:
Refactor tool list item templates for improved URL handling

- Updated _tool_categorized_list_item.html, _tool_grid_list_item.html, and _tool_vertical_list_item.html to conditionally render tool URLs, allowing for clickable rows and cards.
- Simplified the display logic for tool names and descriptions, ensuring consistent styling and behavior across different list formats.
- Enhanced accessibility by adding cursor pointer styles for clickable elements.

julioest committed:
Adds a "Tools" link next to the category dropdown that scrolls users to the Tools section. Includes tooltip on hover.

julioest committed:
Remove "Tools" link and associated tooltip from library preferences template

gregjkal commented on libraries/utils.py:
```diff
@@ -370,3 +371,41 @@ def generate_release_report_filename(version_slug: str, published_format: bool =
+                    f"https://www.boost.org/doc/libs/{version_slug}/{url_path}"
+                )
+            else:
+                tool_dict["url"] = ""
```
```suggestion
        if version and tool_dict.get("version_specific"):
            version_slug = version.stripped_boost_url_slug
            tool_dict["url"] = (
                f"https://www.boost.org/doc/libs/{version_slug}/{url_path}"
            )
            else:
                tool_dict["url"] = ""
```

julioest committed:
Remove "Wave" tool entry from the TOOLS constant in constants.py

gregjkal reviewed (changes requested):
Looking good - couple of questions/suggested changes.

julioest committed:
Update libraries/constants.py

Co-authored-by: Greg Kaleka <greg@gregkaleka.com>

julioest committed:
Update libraries/utils.py

Co-authored-by: Greg Kaleka <greg@gregkaleka.com>

julioest committed:
Guard get_tools against version=None

The version_slug hoist computed version.stripped_boost_url_slug before checking whether version was set, so calling get_tools() with no version threw AttributeError. The library list view does exactly that on some paths. Only compute the slug when a version is actually present.

julioest committed:
Make tool rows real links for accessibility

The tool list rows used an onclick on the whole <tr> to open the doc URL, which keyboard users cannot focus or activate and which had no copyable link. Render the tool name as a normal anchor instead, matching how the library rows and grid cards already work. Addresses the CodeRabbit accessibility findings.

julioest commented on libraries/constants.py:
```diff
@@ -369,3 +369,60 @@
+        "slug": "build",
+        "description": "The Boost build system, including the full Boost version of the jam sources.",
+        "version_specific": True,
+        "url_path": "tools/build/doc/html/index.html",
```
Yup, I tested all of these against production. They return 200 and track the selected version correctly (checked on 1.90, 1.88, and latest).

Left the leading slash off on purpose since `get_tools()` builds the path including that slash

Heads up: if you test locally, it will 404 since those `/doc/libs/...` pages are served from S3 and local doesn't have the content. The URLs are correct, they just resolve on prod/staging.

gregjkal commented on libraries/constants.py:
```diff
@@ -369,3 +369,60 @@
+    {
+        "name": "QuickBook",
+        "description": "QuickBook is a WikiWiki style documentation tool geared towards C++ documentation using simple rules and markup for simple formatting tasks. QuickBook generates BoostBook XML.",
+        "url_path": "/doc/libs/latest/doc/html/quickbook.html",
```
Marked `version_specific: False` but hardcodes `latest`. Viewing 1.88, the other doc tools track 1.88 but QuickBook jumps to latest. Should be `version_specific: True` with `url_path: "doc/html/quickbook.html"`.

gregjkal commented on libraries/constants.py:
```diff
@@ -369,3 +369,60 @@
+    {
+        "name": "Regression",
+        "description": "The Boost regression testing system reporting sources.",
+        "url_path": "https://www.boost.org/doc/contributor-guide/testing/regression-tests.html",
```
Use root-relative `/doc/contributor-guide/testing/regression-tests.html`.

gregjkal commented on templates/libraries/grid_list.html:
```diff
@@ -20,6 +20,18 @@
     {# end libraries list #}
 
+    {# Tools section #}
+    <div id="tools" class="mt-8 mb-5">
```
Tools render even when a category filter is active (`get_queryset` filters libraries, this section doesn't), and repeat on every paginated page. Suggest gating on `{% if not category %}` and last/only page. Same applies to vertical_list.html.

gregjkal commented on templates/libraries/_tool_categorized_list_item.html:
```diff
@@ -0,0 +1,23 @@
+<tr class="border-0 md:border border-gray-200/10 border-dotted md:border-t-0 md:border-r-0 md:border-l-0 md:border-b-1 hover:bg-gray-100 dark:hover:bg-gray-700 transition-all duration-200 ease-in-out">
+  <td class="py-2 align-top md:w-1/5">
+    {% if tool.url %}
+      <a class="mr-1 font-bold capitalize text-sky-600 dark:text-sky-300 hover:text-orange dark:hover:text-orange" href="{{ tool.url }}" target="_blank" rel="noopener noreferrer">{{ tool.name }}</a>
```
`capitalize` renders bcp as "Bcp" and boostlook as "Boostlook". Library rows get away with it because names are already capitalized; drop it here (also in the vertical item template).

gregjkal reviewed (changes requested):
A few more tweaks.

julioest committed:
Fix tool URLs and tools section rendering

QuickBook now tracks the selected version instead of pinning to latest, and Regression uses a root-relative URL so it resolves off prod. Dropped capitalize so bcp/boostlook render correctly, and gated the tools section so it hides under a category filter and only shows on the last page.
